ONSdigital/eq-author-app
# Ear 235 openclose all

> BEFORE MAKING YOUR PR... There must be no linting errors and all tests must pass

### What is the context of this PR?

~~Allows all the section accordions to be opened and closed with a single click.~~

~~The current method involves drilling props down a few layers which works, but I wish the context method would work.~~

~~I did try to use useContext to pass the props down the several layers. But the tests kept breaking even when wrapped in the context provider...~~

So this has been reworked ever so slightly since everyone looked at it last. Basically I noticed there were significant performance issues when you started to add loads of sections. I think they were being caused by lots of rerenders when the open/close state of the accordion was updated.  Which would cause the whole nav to render again including the section-title.

There are several things I've done to address these performance issues:
- Only update the label of the button if one of the accordions is closed
- Removed the useEffect at the top level -  could be done in an event handler instead
- Wrapped a few Callback hooks around certain functions in various components. The callback hook only instantiates a new function if one of the dependencies change in the callback hook. Wrapping this around the SectionTitle helped minimize the amount of stuff being rendered each time

### How to review 
0.5 Open chrome dev tools and click this button in the top left
![Screenshot 2020-06-11 at 12 06 54](https://user-images.githubusercontent.com/22731314/84428855-0bf54680-abdc-11ea-8285-1f0aa116f07e.png)

It should open up tabs on the left like this
![Screenshot 2020-06-11 at 12 07 21](https://user-images.githubusercontent.com/22731314/84428894-1c0d2600-abdc-11ea-896a-9c64d3750de7.png)

The verbose tab at the bottom shows when certain actions are taking longer than expected

1. Open questionnaire
2. Create as many sections as you want but more than 1 (Try like 20 if you can)
3. Hit Open all
    - should change to close all
    - if individual section is opened, they should all change to the expected state instead of the opposite of the current state 
4. Adding a new section and see if that works
5. When all sections are open, the button should say 'Close all'
6. When one section is open, the button should say 'Open all'
7. When there are a lot of sections, you should be able to scroll down the nav and the open all button should be fixed under the add content button.
